crates/bifrost/src/providers/replicated_loglet/replication/checker.rs

```diff
@@ -39,327 +98,747 @@ use restate_types::PlainNodeId;
+        self.per_attribute_counter
+            .get(attr)
+            .map(|counters| counters.num_authoritative_nodes > 0)
+            .unwrap_or(false)
+    }
+
+    fn non_empty_nodes(&self) -> u32 {
+        self.num_nodes
+    }
+
+    /// Is the entire domain empty?
+    fn is_empty(&self) -> bool {
+        self.num_nodes == 0
+    }
 }
 
-impl FMajorityResult {
-    pub fn passed(&self) -> bool {
-        matches!(
-            self,
-            FMajorityResult::Success | FMajorityResult::SuccessWithRisk
-        )
+impl<Attr> Default for FailureDomainState<Attr> {
+    fn default() -> Self {
+        Self {
+            num_nodes: 0,
+            per_attribute_counter: Default::default(),
+        }
+    }
+}
+
+/// A wrapper type that to implement Hash and Eq on a raw pointer allocation, equality
+/// is based on the underlying allocation pointer.
+#[derive(Debug, Clone)]
+struct HashPtr<Attr>(*mut FailureDomainState<Attr>);
+
+// **Safety**: HashPtr is safe to send as it's strictly private to NodeSetChecker and is tied to
+// a heap-allocation that's also self-owned.
+unsafe impl<Attr: Send> Send for HashPtr<Attr> {}
+
+impl<Attr> HashPtr<Attr> {
+    fn as_ref(&self) -> &FailureDomainState<Attr> {
+        // Safety: We are the only one who can create HashingPtr and we guarantee that the
+        // underlying allocation is valid.
+        unsafe { &*self.0 }
+    }
+    fn as_mut_ref(&mut self) -> &mut FailureDomainState<Attr> {
+        // Safety: We are the only one who can create HashingPtr and we guarantee that the
+        // underlying allocation is valid.
+        unsafe { &mut *self.0 }
+    }
+}
+
+impl<Attr> Hash for HashPtr<Attr> {
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        self.0.hash(hasher)
+    }
+}
+
+impl<Attr> Eq for HashPtr<Attr> {}
+impl<Attr> PartialEq for HashPtr<Attr> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl<Attr> HashPtr<Attr> {
+    fn new(value: &mut Box<FailureDomainState<Attr>>) -> Self {
+        // Critical: We want the address of the underlying allocation, not the box itself so the
+        // double dereferencing is essential.
+        //
+        // We could have made new() accept &mut FailureDomainState<Attr> but it would have made it
+        // easy to pass non-boxed references which will move.
+        let value: *mut _ = &mut **value;
+        Self(value)
+    }
+}
+
+#[derive(Debug)]
+struct LocationScopeState<Attr> {
+    /// domains at that scope
+    failure_domains: HashMap<SmartString, Box<FailureDomainState<Attr>>, Xxh3Builder>,
+    /// maps between node-id to the domain it belongs to for fast lookups
+    node_to_fd: HashMap<PlainNodeId, HashPtr<Attr>, Xxh3Builder>,
+    /// replication factor at that scope
+    replication_factor: u8,
+    /// For each attribute, the number of domains at that scope that have all their nodes with the
+    /// attribute.
+    num_complete_domains: HashMap<Attr, u32>,
```

Comment:
For what is this field used? I can only see that we update it but we seem to never really read it.